### PR TITLE
Mirrors + Portals

### PR DESCRIPTION
**Pull request by HifiExperiments**
resurrecting https://github.com/overte-org/overte/pull/44

technique overview:

- extremely naive, does a full offscreen render pass per mirror and then renders the mirror object to sample from the output texture. in the future, this can hopefully be improved to use the stencil buffer without offscreen rendering passes, but we use the stencil buffer for other things so I wasn't able to get that immediately working.
- doesn't currently do anything smart to avoid rendering objects in the mirror view that aren't visible through the mirror...it's possible we'll need a scissor rect/different culling view matrix for this
- currently hardcoded to support 3 levels of mirror recursion and 3 mirrors per depth level. if the mirrors can view each other, that's a lot of fullscreen rendering passes if they are all active, so it will definitely have a perf impact.
- projection matrix is NOT flipped left/right.  instead, we flip the output texture in `ToneMapAndResampleTask.cpp`
- works on shapes, text, images, and models.  for models, each different material/mesh part will count as a separate mirror.  for text, the text and the background also count as separate mirrors (unless one or the other is invisible).

for later work:
- other entity types can be added as followups but aren't as straightforward
- create side should hide properties on types that don't support them
- could completely replace mirror script with mirror prop + renderWithZones

Testing:
need to verify that the following work properly inside of mirrors: VR/stereo?, forward/deferred rendering (toggling should also affect mirrors), rendering effects (shadows, fog, lights), billboarding, cauterization

### CODE REVIEW

File: libraries/shared/src/MirrorMode.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "MirrorMode.h"
+
+const char* MirrorModeNames[] = {
+    "none",
+    "mirror",
+    "portal"
+};
+
+static const size_t MIRROR_MODE_NAMES = (sizeof(MirrorModeNames) / sizeof(MirrorModeNames[0]));
```
[the remark this answers is not in the document]
Comment (HifiExperiments) [reply libraries/shared/src/MirrorMode.cpp:17]: good question, yeah it’s the size of the pointer.  this is a common way of finding the length of a C-style array.  we do the same thing in other files like BillboardMode but we could consider modernizing them in the future for better readability

**Review by ksuprynowicz — APPROVED** [review]

**Comment by ksuprynowicz** [comment]
I tested it in VR and mirrors look amazing :)
Stereo views work correctly for them, and spectator camera works at the same time too. There are a few issues though. The most important one is that camera position for mirrors is delayed by one frame (?) and this breaks the illusion. Or maybe the camera positions are updated only during game loop and not in render loop?
Another issue is that portals didn't seem to be available in the mirror property dropdown.

**Comment by HifiExperiments** [comment]
I filed an issue for the dropdowns getting cut off since it's not specific to this PR: https://github.com/overte-org/overte/issues/750

after digging a little I think the delay is probably due to the "camera correction" code.  after the render thread submits a frame, the present thread applies the correction to account for any camera movement in between, so that the camera in VR is smoother.  but if the correction is to move to the left, it actually needs to move to the *right* in the mirror view.

the secondary camera explicitly disables this, with `batch.disableContextViewCorrection()`, but I think we need to actually take it into account for all the mirror views...I think I need to add a new batch command to either specify a mirror level, which would then flip the x axis of the correction matrix, or to specify a function that does all the mirror view calculations based on the view + correction...

File: libraries/gpu-gl-common/src/gpu/gl/GLBackendTransform.cpp [hunk 2]
```diff
@@ -111,7 +111,11 @@ void GLBackend::TransformStageState::preUpdate(size_t commandIndex, const Stereo
         if (_viewIsCamera && (_viewCorrectionEnabled && _correction.correction != glm::mat4())) {
             // FIXME should I switch to using the camera correction buffer in Transform.slf and leave this out?
             Transform result;
-            _view.mult(result, _view, _correction.correctionInverse);
+            glm::mat4 correction = _correction.correctionInverse;
+            if (_mirrorViewCorrection) {
+                correction = glm::scale(glm::mat4(), glm::vec3(-1.0f, 1.0f, 1.0f)) * correction;
```
Comment (HifiExperiments) [inline libraries/gpu-gl-common/src/gpu/gl/GLBackendTransform.cpp:116]: @ksuprynowicz I added this to attempt to fix the delay in VR, but I wasn't able to test it as we don't use the view correction in desktop and I couldn't get the DebugHMDDisplayPlugin working (seems like the projection matrices are weird).

this math is almost certainly wrong as all the matrix inversion is confusing, but the basic idea is that I think we need to flip the correction matrix across the x axis *before* applying it to the view matrix.  could someone with an HMD test this?  and maybe even futz with it if it doesn't work as is 🙏

**Comment by ksuprynowicz** [comment]
Thank you!
I will check it out in VR as soon as I finish work tomorrow.

**Comment by ksuprynowicz** [comment]
In the newest version it looks like problem with properties got worse - I'm not able to switch on a mirror at all now neither in VR nor in Desktop mode:
20231206_195106_MixedReality
Following change in entityProperties fixes this and allows accessing portal option in the dropdown:
 ```
             {
                label: "Billboard Mode",
                type: "dropdown",
                options: {
                    none: "None",
                    yaw: "Yaw",
                    full: "Full"
                },
                propertyID: "billboardMode",
            },
            {
                label: "Mirror Mode",
                type: "dropdown",
                options: {
                    none: "None",
                    mirror: "Mirror",
                    portal: "Portal"
                },
                propertyID: "mirrorMode",
            },
            {
                label: "Portal Exit",
                type: "string",
                propertyID: "portalExitID",
                showPropertyRule: { "mirrorMode": "portal" },
            },
            {
                label: "Render With Zones",
                type: "multipleZonesSelection",
                propertyID: "renderWithZones",
            }
        ]
```
I also tested it in VR and the result is very unexpected:
https://github.com/overte-org/overte/assets/11568651/5f642cf2-383c-41c8-a05b-dfad0c3f3beb

**Comment by HifiExperiments** [comment]
@ksuprynowicz thanks, that's much smarter 😄 

for VR, I pushed a change to reverse the matrix multiplication but I'm just guessing here

**Comment by vegaslon** [comment]
I tested the latest change in VR and the mirror view is attaching to the head (it rotates and moves based on the head movement) when going back to desktop mode the mirror becomes stationary again.

**Comment by HifiExperiments** [comment]
thanks @vegaslon!  I noticed one suspicious thing so I've tried a different fix, although this one makes the lighting in the mirror weird for me in deferred mode even in desktop, but I'd be curious what it does in VR.

**Comment by HifiExperiments** [comment]
I've hacked together a way to test the view correction myself without an HMD and it still appears broken, will keep investigating

**Comment by HifiExperiments** [comment]
@ksuprynowicz @vegaslon ok, this appears to be working for me now with my view correction hack in desktop.  hopefully it works in VR now!

**Comment by ksuprynowicz** [comment]
I just tested the newest version in VR and it works amazingly well with deferred renderer. There's a bit of some sort oerlaid image over the mirror, but it's hard to tell what it is and it's barely noticeable. In forward renderer the image is still lagging like it did previously. In portal mode everything works as well as in mirror mode as long as both portal and portal entry are visible at the same time. In some camera locations image starts lagging again.
I wonder why the lag in mirror mode only happens on forward renderer and works perfectly on deferred?

**Comment by HifiExperiments** [comment]
ok!  I found a dumb bug and cleaned some stuff up and I *think* this is working for reals now

**Comment by ksuprynowicz** [comment]
Awesome! I will test it tomorrow :)

**Comment by vegaslon** [comment]
Looking good in desktop mode, I noticed when you create a portal the view through it is mirrored, is that intended behavior?

**Comment by HifiExperiments** [comment]
@vegaslon you're totally right, I was accidentally treating all portals as mirrors too.  I *think* I've fixed that...sometimes it's confusing to tell what's happening with a bunch of mirrors/portals set up 😆 

To test, if I set up 2 portals facing each other:
```
{
  let entity1 = Entities.addEntity({type: "Box", position: Vec3.sum(MyAvatar.position, {x:0, y: 0, z: -3}), dimensions: {x: 6, y: 6, z: 0.1}, mirrorMode: "portal"})
  let entity2 = Entities.addEntity({type: "Box", position: Vec3.sum(MyAvatar.position, {x:0, y: 0, z: 3}), rotation: {"x":0,"y":1,"z":0,"w":0}, dimensions: {x: 6, y: 6, z: 0.1}, mirrorMode: "portal", portalExitID: entity1 })
  Entities.editEntity(entity1, { portalExitID: entity2 });
}
```
I see my own back 3 times, and if I raise my hand with "x", they all move the same hand.

And if I do the same set up with two mirrors, the results look how I'd expect too.

**Comment by ksuprynowicz** [comment]
I just tested the newest branch in VR and now mirrors work amazingly well both in forward and deferred renderers. With portals there's still a problem. When I rotate my head the image from the portal rotates in an incorrect way. When portals are enabled there is also a large crosshair in the center of the screen. It shows up only in deferred renderer:
20240101_151246_MixedReality

**Comment by ksuprynowicz** [comment]
When I rotate my head in the vertical axis the image from the portal seems "stuck" to VR headset orientation, and if I tilt my head the image tilts at twice the angle.

**Comment by HifiExperiments** [comment]
thanks @ksuprynowicz, *now* I think I've got the portal calculation right for real 🤞 

the crosshair is a debug cursor from some other effect.  is it possible you toggled it on with luci or something?  I forget which effect it is but I don't see it so maybe clearing your settings would get rid of it?

**Comment by ksuprynowicz** [comment]
I tested the newest version in VR again, and now both portals and mirrors work absolutely amazing on both renderers

**Comment by ksuprynowicz** [comment]
I'm sorry for accidentally merging it, I wanted to leave a comment and I pressed wrong button

**Pull request merged**